.github/actions/github-release/main.js

```diff
@@ -11,66 +11,104 @@ function sleep(milliseconds) {
 async function runOnce() {
   // Load all our inputs and env vars. Note that `getInput` reads from `INPUT_*`
   const files = core.getInput('files');
-  const name = core.getInput('name');
   const token = core.getInput('token');
   const slug = process.env.GITHUB_REPOSITORY;
   const owner = slug.split('/')[0];
   const repo = slug.split('/')[1];
   const sha = process.env.GITHUB_SHA;
+  let name = 'dev';
+  if (process.env.GITHUB_REF.startsWith('refs/tags/v')) {
+    name = process.env.GITHUB_REF.substring(10);
+  }
 
   core.info(`files: ${files}`);
   core.info(`name: ${name}`);
   core.info(`token: ${token}`);
 
   const octokit = new github.GitHub(token);
 
-  // Delete the previous release since we can't overwrite one. This may happen
-  // due to retrying an upload or it may happen because we're doing the dev
-  // release.
-  const releases = await octokit.paginate("GET /repos/:owner/:repo/releases", { owner, repo });
-  for (const release of releases) {
-    if (release.tag_name !== name) {
-      continue;
+  // For the `dev` release we may need to update the tag to point to the new
+  // commit on this branch. All other names should already have tags associated
+  // with them.
+  if (name == 'dev') {
+    let tag = null;
+    try {
+      tag = await octokit.request("GET /repos/:owner/:repo/git/refs/tags/:name", { owner, repo, name });
+      core.info(`found existing tag`);
+      console.log("tag: ", JSON.stringify(tag.data, null, 2));
+    } catch (e) {
+      // ignore if this tag doesn't exist
+      core.info(`no existing tag found`);
+    }
+
+    if (tag === null || tag.data.object.sha !== sha) {
+      core.info(`updating existing tag or creating new one`);
+      // Delete the previous release for this tag, if any
+      try {
+        core.info(`fetching release for ${name}`);
+        const release = await octokit.repos.getReleaseByTag({ owner, repo, tag: name });
+        core.info(`deleting release ${release.data.id}`);
+        await octokit.repos.deleteRelease({ owner, repo, release_id: release.data.id });
+      } catch (e) {
+        // ignore, there may not have been a release
+        console.log("ERROR: ", JSON.stringify(e, null, 2));
+      }
+
+      try {
+        core.info(`updating dev tag`);
+        await octokit.git.updateRef({
+            owner,
+            repo,
+            ref: 'tags/dev',
+            sha,
+            force: true,
+        });
+      } catch (e) {
+        console.log("ERROR: ", JSON.stringify(e, null, 2));
+        core.info(`creating dev tag`);
+        try {
+          await octokit.git.createTag({
+            owner,
+            repo,
+            tag: 'dev',
+            message: 'dev release',
+            object: sha,
+            type: 'commit',
+          });
+        } catch (e) {
+          // we might race with others, so assume someone else has created the
```

Comment:
Is it worth distinguishing the "tag already exists because race" case from other error cases? There's a potential that e.g. a network hiccup is taken as a collision and we end up without a `dev` release, otherwise, I think. Perhaps it's not really worth worrying too much about; not sure (the `dev` release is certainly less important than named releases).

Reply:
I don't currently know how to handle the right errors (it feels like it's just a bunch of string matching?), but I figured that it's always possible to inject more JS later to figure this out if it becomes an issue. I don't think it's as easy as like `if (e instanceof TheErrorWeCareAbout)` since I think we'd have to inspect the error which is just the JSON response and figure out the cause of the error from that.